ams/server/src/main/java/com/netease/arctic/server/optimizing/plan/AbstractPartitionPlan.java

```diff
@@ -99,24 +96,19 @@ public long getCost() {
   }
 
   @Override
-  public void addFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
-    evaluator().addFile(dataFile, deletes);
-    if (evaluator().isFragmentFile(dataFile)) {
-      fragmentFiles.put(dataFile, deletes);
-    } else {
-      segmentFiles.put(dataFile, deletes);
+  public boolean addFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+    boolean added = evaluator().addFile(dataFile, deletes);
+    if (!added) {
+      // if the Data file is not added, it's Delete files should be not be removed from iceberg
+      deletes.stream().map(delete -> delete.path().toString()).forEach(protectedDeleteFiles::add);
+      return false;
     }
-    for (IcebergContentFile<?> deleteFile : deletes) {
-      if (deleteFile.content() == FileContent.POSITION_DELETES) {
-        posDeleteFileMap
-            .computeIfAbsent(deleteFile.path().toString(), delete -> Sets.newHashSet())
-            .add(dataFile);
-      } else {
-        equalityDeleteFileMap
-            .computeIfAbsent(deleteFile.path().toString(), delete -> Sets.newHashSet())
-            .add(dataFile);
-      }
+    if (evaluator().fileShouldRewrite(dataFile, deletes)) {
+      rewriteDataFiles.put(dataFile, deletes);
+    } else if (evaluator().segmentFileShouldRewritePos(dataFile, deletes)) {
+      rewritePosDataFiles.put(dataFile, deletes);
     }
```

Comment:
It seems it is possible a data file can not match both conditions here, do we should handle this situation?

Reply:
Considering the implementation of the Evaluator, this scenario should not occur, but it is quite unclear. 

We have two ways to handle this situation, throwing an exception or returning false. I prefer the latter, what do you think?

Reply:
Returning false seems good, but we should put the delete files path into the `reservedDeleteFiles` too.

Reply:
OK, it will.